### PR TITLE
Windows support

### PR DESCRIPTION
**Pull request by ibaned**
This is a work in progress branch to support Windows.

### CODE REVIEW

**Comment by crtrott** [comment]
Other than the clang format thing this looks good so far.

**Comment by ibaned** [comment]
Thanks! I have a couple remaining issues that could use the attention of yourself or @dsunder

**Comment by masterleinad** [comment]
I also need #2664, or more precisely disabling libdl and profiling support, and
```C++
diff --git a/cmake/kokkos_compiler_id.cmake b/cmake/kokkos_compiler_id.cmake
index 1aa93537..166e0125 100644
--- a/cmake/kokkos_compiler_id.cmake
+++ b/cmake/kokkos_compiler_id.cmake
@@ -4,13 +4,14 @@ SET(KOKKOS_CXX_COMPILER ${CMAKE_CXX_COMPILER})
 SET(KOKKOS_CXX_COMPILER_ID ${CMAKE_CXX_COMPILER_ID})
 SET(KOKKOS_CXX_COMPILER_VERSION ${CMAKE_CXX_COMPILER_VERSION})
 
-# Check if the compiler is nvcc (which really means nvcc_wrapper).
-EXECUTE_PROCESS(COMMAND ${CMAKE_CXX_COMPILER} --version
-                COMMAND grep nvcc
-                COMMAND wc -l
-                OUTPUT_VARIABLE INTERNAL_HAVE_COMPILER_NVCC
-                OUTPUT_STRIP_TRAILING_WHITESPACE)
-
+IF(NOT CMAKE_CXX_COMPILER_ID MATCHES "MSVC")
+  # Check if the compiler is nvcc (which really means nvcc_wrapper).
+  EXECUTE_PROCESS(COMMAND ${CMAKE_CXX_COMPILER} --version
+                  COMMAND grep nvcc
+                  COMMAND wc -l
+                  OUTPUT_VARIABLE INTERNAL_HAVE_COMPILER_NVCC
+                  OUTPUT_STRIP_TRAILING_WHITESPACE)
+ENDIF()
 
 STRING(REGEX REPLACE "^ +" ""
        INTERNAL_HAVE_COMPILER_NVCC "${INTERNAL_HAVE_COMPILER_NVCC}")
```
here. After that, I get roughly to https://ci.appveyor.com/project/masterleinad/kokkos/builds/30271043.

**Comment by ibaned** [comment]
GCC-4.8.4 branch: `unexpected EOF`... not sure what that means in the PR system.

Clang builds fail with:
```
view_layoutstride_left_to_layoutright_assignment: srand(1579627206)
1392/home/travis/build/kokkos/kokkos/core/unit_test/TestViewLayoutStrideAssignment.hpp:829: Failure
1393Death test: { dst = src; }
1394    Result: died but not with expected error.
1395  Expected: View assignment must have compatible layouts
1396Actual msg:
1397[  DEATH   ] 
1398[  FAILED  ] serial.view_layoutstride_left_to_layoutright_assignment (115 ms)
```

**Comment by masterleinad** [comment]
> GCC-4.8.4 branch: `unexpected EOF`... not sure what that means in the PR system.

This is a problem with the machine the machine the check was run on and likely not with this pull request.

Retest this please.

**Comment by masterleinad** [comment]
https://ci.appveyor.com/project/masterleinad/kokkos/builds/30294952

**Comment by masterleinad** [comment]
I pushed a configuration file for appveyor. This is the current build: https://ci.appveyor.com/project/masterleinad/kokkos/builds/30344370

**Comment by masterleinad** [comment]
From my point of view, this is good enough for a first step. In particular, compiling all the examples with the current state works but there are some failing tests, see https://ci.appveyor.com/project/masterleinad/kokkos/builds/30980737. Furthermore, we can use `appveyor` for CI.
I am happy to clean this up here or in a separate pull request if others agree.

**Comment by crtrott** [comment]
What the hell is happening here?

**Comment by masterleinad** [comment]
> What the hell is happening here?

Rebasing was a bit challenging here. Should be good now (https://ci.appveyor.com/project/masterleinad/kokkos/builds/31031585).

File: core/src/impl/Kokkos_Atomic_Load.hpp [hunk 1]
```diff
@@ -168,6 +168,15 @@ inline T _atomic_load(T* ptr, MemoryOrder) {
   return *ptr;
 }
 
+#elif defined(KOKKOS_ENABLE_WINDOWS_ATOMICS)
+
+template <class T, class MemoryOrder>
+inline T _atomic_load(T* ptr, MemoryOrder) {
+  Kokkos::memory_fence();
+  return *ptr;
```
Comment (crtrott) [inline core/src/impl/Kokkos_Atomic_Load.hpp]: This doesn’t make any sense. Probably you want to create a temporary andcreturn after second memory fence? But even then that is not correct for example for complex ...

File: core/src/impl/Kokkos_Atomic_Store.hpp [hunk 2]
```diff
@@ -164,6 +164,15 @@ inline void _atomic_store(T* ptr, T val, MemoryOrder) {
   *ptr = val;
 }
 
+#elif defined(KOKKOS_ENABLE_WINDOWS_ATOMICS)
+
+template <class T, class MemoryOrder>
+inline void _atomic_store(T* ptr, T val, MemoryOrder) {
+  Kokkos::memory_fence();
+  *ptr = val;
```
Comment (crtrott) [inline core/src/impl/Kokkos_Atomic_Store.hpp]: Only works for word size T

File: core/src/impl/Kokkos_Atomic_Load.hpp (hunk 1 above)
Comment (masterleinad) [reply core/src/impl/Kokkos_Atomic_Load.hpp]: I am clearly out of my comfort zone here with `atomics` on Windows. If you have any good suggestions on what to do here, I am happy to adapt.

I mostly want to make sure that we can check that compiling with MSVC continues to work (via enabling appveyor for all PRs). From experience, it is a real struggle to keep that compiler happy if we don't check regularly. I would be fine with just aborting for cases that didn't work before and fix functionality and tests later.

**Comment by masterleinad** [comment]
Current build status: https://ci.appveyor.com/project/masterleinad/kokkos/builds/31238960
and tests: https://ci.appveyor.com/project/masterleinad/kokkos/builds/31240932

**Comment by masterleinad** [comment]
OK. I will clean up the history as well.

**Comment by masterleinad** [comment]
@crtrott I marked the tests that currently don't compile accordingly.

File: containers/unit_tests/TestErrorReporter.hpp [hunk 3]
```diff
@@ -50,6 +50,10 @@
 #include <Kokkos_Core.hpp>
 #include <Kokkos_ErrorReporter.hpp>
 
+#ifndef M_PI
+#define M_PI 3.14159265358979323846
+#endif
```
Comment (dalg24) [inline containers/unit_tests/TestErrorReporter.hpp]: Did you try
```C++
#define _USE_MATH_DEFINES
#include <cmath>
```

Comment (masterleinad) [reply containers/unit_tests/TestErrorReporter.hpp]: I think I did. The problem is if `<cmath>` is included somewhere else first, `M_PI` is still not defined.

**Comment by crtrott** [comment]
Hi,

so I just got a new computer at home which has enough space on C to actually install visual studio :-) The checkout and build worked so that is awesome. But the tests fail. I am looking into that now a bit.

**Comment by masterleinad** [comment]
Yes, most of the failing tests fail with
```
Kokkos::Impl::SharedAllocationRecord 'v' failed decrement count = -1
```
I haven't investigated, yet.

**Comment by crtrott** [comment]
I fixed that already. Working on more atomic fixes.

**Comment by crtrott** [comment]
Ok a lot of the trouble is coming from unsigned long not being 64 bit on Windows …
Potentially we need to replace all unsigned long (and long) with uint64_t and int64_t respectively ....

**Comment by crtrott** [comment]
Args didn't really want to push here, but rather do this with a PR from my fork against Ibaned … but apparently didn't set tracking mode correct in visual studio …

Ok my atomic changes might have broken Linux builds since I moved some generic versions around. We need to test that (but I guess the pr testing will do so). Also the format is probably wrong.

**Comment by crtrott** [comment]
Ok I think I got it now. This also passes all tests on Windows with Serial enabled. (At least in Release mode, in debug mode its still working on KokkosCore_PerfTestExec ...)

**Comment by crtrott** [comment]
@masterleinad feel free to rewrite history now to clean that up.

**Comment by masterleinad** [comment]
> @masterleinad feel free to rewrite history now to clean that up.

Will do and also enable running the tests again.

**Comment by masterleinad** [comment]
@crtrott I cleaned the history a little bit and let `appveyor` also run the tests except for `KokkosCore_PerfTestExec`. For me, this is good enough now and I don't intend to do anything else here.

**Comment by crtrott** [comment]
FYI: this did pass spot check on one of our machines:
```c++
Repository Status:  ce9a339697a4930d1bd8534ba13db17addc88813 Windows: Fix Serial only build on Linux


Going to test compilers:  gcc/7.3.0 gcc/8.3.0 gcc/9.1 intel/18.0.5 clang/8.0 cuda/10.1
Testing compiler gcc/7.3.0
  Starting job gcc-7.3.0-OpenMP-release
kokkos options: 
kokkos devices: OpenMP
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized 
  PASSED gcc-7.3.0-OpenMP-release
Testing compiler gcc/8.3.0
  Starting job gcc-7.3.0-Pthread-release
kokkos options: 
kokkos devices: Pthread
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized 
  PASSED gcc-7.3.0-Pthread-release
Testing compiler gcc/9.1
  Starting job gcc-8.3.0-OpenMP-release
kokkos options: 
kokkos devices: OpenMP
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized 
  PASSED gcc-8.3.0-OpenMP-release
  Starting job gcc-9.1-OpenMP-release
kokkos options: 
kokkos devices: OpenMP
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized 
  PASSED gcc-9.1-OpenMP-release
Testing compiler intel/18.0.5
  Starting job gcc-9.1-Serial-release
kokkos options: 
kokkos devices: Serial
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized 
  PASSED gcc-9.1-Serial-release
Testing compiler clang/8.0
  Starting job intel-18.0.5-OpenMP-release
kokkos options: 
kokkos devices: OpenMP
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized 
  PASSED intel-18.0.5-OpenMP-release
  Starting job clang-8.0-Cuda_OpenMP-release
kokkos options: 
kokkos devices: Cuda,OpenMP
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized 
  PASSED clang-8.0-Cuda_OpenMP-release
Testing compiler cuda/10.1
  Starting job clang-8.0-Pthread_Serial-release
kokkos options: 
kokkos devices: Pthread,Serial
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized 
  PASSED clang-8.0-Pthread_Serial-release
  Starting job cuda-10.1-Cuda_OpenMP-release
kokkos options: 
kokkos devices: Cuda,OpenMP
kokkos cxx: -O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized 
  PASSED cuda-10.1-Cuda_OpenMP-release
#######################################################
PASSED TESTS
#######################################################
clang-8.0-Cuda_OpenMP-release build_time=69 run_time=222
clang-8.0-Pthread_Serial-release build_time=42 run_time=474
cuda-10.1-Cuda_OpenMP-release build_time=106 run_time=210
gcc-7.3.0-OpenMP-release build_time=34 run_time=159
gcc-7.3.0-Pthread-release build_time=31 run_time=428
gcc-8.3.0-OpenMP-release build_time=36 run_time=149
gcc-9.1-OpenMP-release build_time=40 run_time=149
gcc-9.1-Serial-release build_time=35 run_time=223
intel-18.0.5-OpenMP-release build_time=107 run_time=160
```

**Comment by crtrott** [comment]
I will do some more checks with older compilers and XL but other than that I am good to merge this in.

**Comment by crtrott** [comment]
Ok this is good to go. Also my older compiler checks passed. Only HPX failed the incremental tests, but it does so on develop too.

**Pull request merged**